No longer manually starting / ending run loop for Stripe request.

johnnyoshika opened the pull request:
Using Ember.run.begin() before making the Stripe request prevents the UI to update with any pending changes until the Ember.run.end() is called.  This is problematic in case you want to show a spinner while the Stripe request is happening.

Here's a jsbin that demonstrates this:
http://jsbin.com/qacava/1/edit?html,js,output
  * clicking the `Without Run Loop` button will display proper messages while running a long task
  * clicking the `With Run Loop` button will suppress message changes

What was the reason for manually starting the Ember loop?

It seems to work fine (both resolve and reject) without the manual Ember loop.

buritica commented:
hey @johnnyoshika, thanks for this. The reason to use the Ember loop manual was because in slow connections the promises were resolving before the requests were happening, this was also several Ember versions ago so this may not be an issue any more. I'll run some tests to make sure things are good in our implementations and if all looks good I'll merge.

Thanks again,

johnnyoshika commented:
Thanks @buritica! I look forward to your finding.  If you merge this into source, I'll delete my fork of your repository and use yours directly again.

Cheers!

buritica commented:
hey @johnnyoshika unfortunately just by switching out your branch and linking with npm breaks 4 of my acceptance tests. I'm going to have to dig deeper into this, but unfortunately I can't merge this yet. I'm going to try to replicate those scenarios in tests here and see what happens.

I'll keep the PR open until I have some solid findings, thanks again.

buritica commented:
Hey @johnnyoshika, after much research here are my conclusions:
- it is the [recommended approach](https://github.com/eoinkelly/ember-get-what-from-where/blob/master/ember-run-loop.md#how-do-i-wrap-code-in-a-run-loop) to use Ember.run to wrap any asynchronous code.
- as much as I wanted the higher-level `Ember.run()` to function, I believe the way Stripe deals with it's ajax requests (they use a private iframe) prevents Ember from detecting what is going on in the background. This means that I must use the lower-level manual `.begin` and `.end` to abide by the run loop.
- even though your jsbin shows the UI is not updated, I think that is the way it is supposed to work for long tasks. The run loop is designed to ensure `"any deferred actions including bindings and views updates are flushed at the end."`
- I understand your use case, so I'm open to a counter proposal, but I believe removing the run loop is not the way to proceed. I could consider emitting custom events to inform about the life-cycle of the request? Haven't done this before, but I think it may be a more elegant solution than getting rid of the async flow control which could have unintended consequences.

Please let me know your thoughts, for now I'm closing this PR.

johnnyoshika commented:
@buritica, that's very interesting. Thank you for the research.  I must admit that I'm not all that familiar with the run loop, but your logic makes sense.  To get around my problem, what do you think about wrapping my code that updates the UI in a run loop so that the UI updates before your stripe service is called?  I modified the jsfiddle to demonstrate:

I added this:
```
      Ember.run.begin();
      this.modelFor(this.routeName).set('message', 'start');
      Ember.run.end();
      // now call taskWithRunLoop
```

http://jsbin.com/sitegiwoxi/2/edit

buritica commented:
I think rather than you calling the runloop manually, you can schedule the Stripe.createToken (or any task with a run loop) to happen `afterRender`. Have a look at the third task I added to your JSbin http://jsbin.com/nofoviwiho/1/edit

johnnyoshika commented:
Excellent!  Thank you for your help.  That's what I'll do.

johnnyoshika commented:
Interesting.  I just upgrade to your latest release (v2.0) and without making any code changes, the UI change is no longer being postponed.  This worked even without implementing your `afterRender` technique.  I see that you're still using Ember.run.begin() and Ember.run.end(). Did anything change?

buritica commented:
Interesting. Nothing really changed, we are no longer injecting ourselves and rather using new 1.10 service injection. 




We also removed ember-data deps. Thats it. 



--

Juan Pablo Buritica

VP of Engineering - Ride

t: @buritica

On Mon, Mar 23, 2015 at 10:32 AM, johnnyoshika <notifications@github.com>
wrote:

> Interesting.  I just upgrade to your latest release (v2.0) and without making any code changes, the UI change is no longer being postponed.  This worked even without implementing your `afterRender` technique.  I see that you're still using Ember.run.begin() and Ember.run.end(). Did anything change?
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/ride/ember-stripe-service/pull/10#issuecomment-85026399

johnnyoshika commented:
I stand to be corrected.  I don't know why it was working as I previously described.  Maybe some old caching?  Anyways, I do indeed need the `afterRender` technique after all.
